### ISSUE TITLE
InterMine Core Data model - Annotation and assembly

### ISSUE THREAD

**Issue by julie-sullivan** [intermine/intermine#1596]
Problem: Genomic data for the same taxon ID have different annotation and assembly versions.
 
Solution: Add annotation and assembly versions to the Organism object in the core InterMine data model.

```
  <class name="Organism" is-interface="true">
    ... 
    <attribute name="annotationVersion" type="java.lang.String"/>
    <attribute name="assemblyVersion" type="java.lang.String"/>
  </class>
```

We use the organism name for querying, so the user interface should be fine. Even if a mine only has one version, this is still useful information. 

What are the other dangers? Do we use taxonID as a key in widgets? Will need to update to use organism name instead.

**Comment by julie-sullivan** [comment]
They would be differentiated by the organism name, even though they have the same taxon ID:

```
	B. distachyon
	B. distachyon Bd21-3 early-release
```

**Database build**

- Integration keys set to be organism name (not taxon ID)
- Annotation data assigned to correct feature with correct organism

**UI**

- Templates use organism name
- Update widgets to use organism name (if this isn't the case)

**Comment by julie-sullivan** [comment]
From @sammyjava 
```
So I'd like Organism.variety added .. or call it subspecies or germplasm or landrace or whatever you want. I liked "variety" as a general catch-all. Specifically this was necessitated by Chickpea desi (the green or dark kind) and Chickpea kabuli (the tan kind, AKA garbanzo).
```

**Comment by sammyjava** [comment]
Actually, on second thought, annotation and assembly version don't really belong on Organism - they refer to the same organism, just different DNA analysis. My thing, variety, IS actually the organism. In principle, we **should have a class called Genome** which has .organism, .annotationVersion and .assemblyVersion. That's what it's usually called, a particular genome assembly and annotation for the organism. A nice example is Maize B73_v1, B73_v2, B73_v3 and B73_v4. All are Maize B73 (so variety=B73) but four different genome assemblies.

**Comment by julie-sullivan** [comment]
From @rachellyne:

Easiest scenario - everything is in the organism name field: `organism_species_genomeAssembly_GenomeAnnotation`.  If you have the species and/or assembly and/or annotation in different fields it will be complicated whether they are under the Organism class or a Genome class.

This would only be a problem if you want to do a more general query over everything (all data from a organism?).  Would you ever want to do this?

**Comment by julie-sullivan** [comment]
@sammyjava I agree your way is more correct but it's adding a bit of complexity. Do we think that complexity is adding enough to the user experience to justify? Are people going to be confused? Or is it going to help?

1. we want to be future proof (as possible!). I don't want to have to do this again
2. some mines don't have strains or genome versions.
3. ease of use. can users easily without help find what they are looking for

**Comment by rachellyne** [comment]
Species - I actually mean strain, variety or equivalent.

**Comment by sammyjava** [comment]
I don't think adding attributes adds complexity. Adding references and collections, perhaps. Plus, isn't honesty important? The 
genus is the genus. The species is the species. The taxon ID is the taxon ID. There are addditional things that specify a particular 
genome. Like variety, etc. So I think it just clusterifies things to munge the genus and species names.

**Comment by julie-sullivan** [comment]
**PROPOSED SOLUTION**

```
  <class name="Organism" is-interface="true">
    ... 
    <attribute name="annotationVersion" type="java.lang.String"/>
    <attribute name="assemblyVersion" type="java.lang.String"/>
    <attribute name="variety" type="java.lang.String"/>
  </class>
```

This does require that the `Organism.name` be unique, instead of the `Organism.taxonId`.

@sammyjava @JoeCarlson You said that your changes are already on github, where you had to change the key from taxon ID to organism name. Can you point out those changes? I'd like to get them in the core InterMine code.

**Comment by sammyjava** [comment]
I don't see why Organism.name has to be unique. I use Organism.taxonId,Organism.variety as the merge key. Here's my stuff:
https://github.com/LegumeFederation/intermine-bio-core

**Comment by julie-sullivan** [comment]
@sammyjava So do you have duplicates in your organism drop downs? In widgets and on templates and query results? Joe avoids this by making the names descriptive, e.g. `Brachypodium distachyon Bd21-3` and `Brachypodium distachyon` (same taxon)

You don't have to follow this pattern, but I don't see how your users are going to be able to differentiate the organisms if you don't.

**Comment by julie-sullivan** [comment]
@sammyjava Thanks for the github link, super useful!!

**Comment by sammyjava** [comment]
@julie-sullivan Perhaps the organism drop-downs have to be modified to accommodate Organism.variety. I don't see a reason to design the database against an existing sub-functional UI. Take a look at ChickpeaMine. I'm sure there are some issues in places, but I didn't change the UI at all. It mostly works fine. ChickpeaMine is an existing example of a mine that uses Organism.variety.

**Comment by JoeCarlson** [comment]
I think that an issue here is that what works for one mine will not be universal. And that’s a good thing.

I needed to replace taxonId as the primary key in the organism table. This was for 2 reasons: first, it’s not unique in our mine so this is obvious. The second reason is that we’re also using intermine as a middleware layer between the db and our main web app. The main web app uses a different primary key (we call it proteomeId) for the annotation. I’ve incorporated this into my organism data model. so that our web app can make calls to intermine using that to specify the organism. What I’d like to see is to make the organism primary key be something we can specify in a properties file.

Since the organism short name is used to populate menus, I needed to make it easy to distinguish different versions of the same genus and species by adding either a strain or release status (the term we use is ‘early-release’ for data not yet incorporated into family builds) to make the distinction clear. Organism.name is not unique for me.

One minor quibble I have about Julie’s addition of variety to the organism model is that this is this term is common, but not universal. Identifiers narrower than species for food crops are ‘cultivars’. For organisms collected in the wild, we just use ‘lines’, and organisms collected in the wild but propagated in a lab are ‘accessions’. It’s great to have variety in there as a field but it may be supplanted in other mines depending on their needs. (And I have no clue of how hybrids are named.) My concern is that there will be code that assumes the combination of genus, species and variety are unique. Make it configurable so that if an assumption is made, the mine-setter-upper can specify that genus, species and cultivar give a unique row.

Thanks for the discussion,

Joe

**Comment by sammyjava** [comment]
For sure. I chose "variety" in my mines as a catch-all for: germplasm, landrace, cultivar, accession, strain, stock, and many others, all of which represent variations of a particular species. It's not comprehensive, but fits what we need - ability to load multiple organisms that are of the same species. @JoeCarlson

**Comment by julie-sullivan** [comment]
On the call we decided that adding `variety` to organism is not the most biologically correct, it's the best model for our (querying) purposes. 

We didn't agree on a name at all though, and someone (Jennifer?) suggested we take a vote.

**Comment by julie-sullivan** [comment]
“Variety” - Generic enough? Does this cover all our use cases.

Other suggestions
- Strain
- Subspecies
- Stock
- Line 
- Accession
- Subtype
- Ecotype
- Isolate
- Others? ...

**Comment by sammyjava** [comment]
I still like it. I think Vivek is using it now as well, at least he had me put my few core code mods up on GitHub so he could use them.  It's the only word I could come up with that doesn't carry specific meaning like all the others do, at least with certain communities. Plus, I started with cowpea desi and kabuli, which are pretty generic varieties themselves, having lots of variability within them.

**Comment by JoelRichardson** [comment]
I still think this is not a good idea. Alternatives have not been
sufficiently discussed, nor have the implications of mooshing this concept
(along with assembly version) all into Organism. It seems to me that the
proposed changes, if enacted, should be accompanied by changing the class
name from Organism to GenomeAssembly (or something).

My $0.02

**Comment by vivekkrish** [comment]
Hello Everyone -

Sam is right, we did start using the simplistic extension to the **Organism** class for MedicMine (see commits: https://github.com/jcvi-plant-genomics/intermine/compare/stable...jcvi-plant-genomics:feature/sammyjava/organism-variety).

However, after the last InterMine Community Call which discussed the model changes, I went back to further investigate the implications of this change, both in the scope of ThaleMine and MedicMine and came to the realization that it might be too simplistic and probably not well suited for our needs (contrary to the original thoughts I expressed in the call).

It would seem that the addition of a new class, such as **Strain** (suggested by Joel and others in the call), which is referenced by the **Organism** (and likely via a reverse relationship as well) might be a better approach. Here are my thoughts:

- **SequenceFeature** entities would continue to reference **Organism**. 
- **Organism** would reference one or more **Strain** entities, each of which have additional attributes/references/collections associated with them.

Here is how it could potentially be modeled:
```
<class name="Organism" is-interface="true">
    <attribute name="taxonId" type="java.lang.Integer"/>
    <attribute name="shortName" type="java.lang.String"/>
    <attribute name="commonName" type="java.lang.String"/>
    <attribute name="genus" type="java.lang.String"/>
    <attribute name="species" type="java.lang.String"/>
    <attribute name="name" type="java.lang.String"/>
    <collection name="strains" referenced-type="Strain" reverse-reference="organism" />
</class>

<class name="Strain" extends="BioEntity" is-interface="true">
    <attribute name="habitat" type="java.lang.String" />
    <attribute name="geoLocation" type="java.lang.String" />
    <reference name="organism" referenced-type="Organism" reverse-reference="strains" />
    <reference name="type" referenced-type="StrainType" reverse-reference="strains"/>
    <collection name="stocks" referenced-type="Stock" reverse-reference="accession" />
    <collection name="backgroundStocks" referenced-type="Stock" reverse-reference="backgroundAccessions" />
</class>

<class name="StrainType" is-interface="true">
    <attribute name="identifier" type="java.lang.String"/>
    <attribute name="name" type="java.lang.String"/>
    <attribute name="description" type="java.lang.String"/>
    <collection name="strains" referenced-type="Strain" reverse-reference="type"/>
</class>
```

The addition of a *Strain* class will ensure that we can reference other entities such as *Stocks* (in particular for Arabidopsis).

And, since *Strain* extends *BioEntity*, attributes such as _primaryIdentifier_, _name_, and collections like _synonym_ can be used to describe the strain's _accessionNumber_, _name_, _aliases_, respectively.

An example Strain entry would be something like:
**primaryIdentifier**: 4735
**name**: An-1
**habitat**: Botanic Garden
**geoLocation**: Antwerpen, Belgium
**StrainType**: Ecotype
**BackgroundStocks**: [ _CS300, CS301, CS302, CS303, ..._ ]

The **StrainType** class can be driven by a controlled vocabulary, supporting several type of species variation:
[ _Strain, Cultivar, Ecotype, Forma, Subforma, Subspecies, Variety, SubVariety, ...._ ]

In order to show distinct **Organism** <--> **Strain** combinations in the user interface (e.g. when picking Organism as a constraint in a query), the _Strain.name_ would need to be appended to the species string (e.g. _A. thaliana An-1_), similar to what Sam has already implemented now (in ChickpeaMine). But, this wouldn't be possible if the **Strain** is a _collection_ within **Organism**. It would have to be converted to a _reference_ instead, and the **Organism** rows would need to be duplicated.

Look forward to discussing this further at the next call, this Thursday. Sorry for throwing a wrench in the mix!

Thank you.

cc @sammyjava @adf-ncgr @agnespychan @cdtown

**Comment by sammyjava** [comment]
This is kind of along the same lines of what I implemented for CowpeaMine, which has a bunch of mapping experiments between accessions or germplasms:
```
  <class name="Germplasm" is-interface="true">
    <attribute name="primaryIdentifier" type="java.lang.String"/>
    <attribute name="secondaryIdentifier" type="java.lang.String"/>
    <attribute name="description" type="java.lang.String"/>
    <attribute name="patentNumber" type="java.lang.String"/>
    <attribute name="url" type="java.lang.String"/>
    <attribute name="country" type="java.lang.String"/>
    <reference name="organism" referenced-type="Organism"/>
  </class>
```
In that case, however, I don't have any genomes. We do have the first sequenced cowpea genome coming out now, but that'll be like the usual case, a single Organism+variety with genomic data. So I've kind of been dealing with this issue differently on the genomic and genetic sides, so far.

**Comment by adf-ncgr** [comment]
@vivekkrish - when we were discussing this before,  you had Organism referencing a single Strain but now you have it as a collection instead; I might be misunderstanding but if Organism remains as representing the "species" and SequenceFeatures just reference it, then aren't we back to not being able to figure out which "strain" a particular gene model or chromosome is representing? I had thought you were earlier proposing a one-to-at-most-one Organism-Strain relationship which does not have this problem.

**Comment by JoelRichardson** [comment]
I agree, but I don't think it would be right to have an organism reference just one (or 0) strains. To me it makes sense that multiple strains/subspecies may all be of the same organism.

**Comment by adf-ncgr** [comment]
@JoelRichardson - to your earlier point I think the current discussion is limited in the scope of its proposed "mooshing" to "infraspecific" identity of genomes and is not intended to pull in the assembly versioning concept (though that had been discussed as a possibility at one point, but I think most people agreed that it wasn't likely that mines would need to maintain concurrent genomic versions).

To your more recent point regarding cardinality of reference from Organism to Strain, I think you're basically saying you want to leave Organism as representing "species" and allow infraspecific stuff to hang off of that as children. Would you feel better about adding an optional Strain reference in SequenceFeature for cases where the infraspecificity of the datum needed to be made explicit? Another possible alternative might be to introduce hierarchy into the Organism table, though that seems like it could be a little harder to deal with in some ways.

**Comment by vivekkrish** [comment]
Hi @adf-ncgr, Yes, actually that is what I was proposing in our earlier discussion. And yes, it does make sense to have a one-to-one relationship between Organism and Strain, in the context of being able to distinguish the origin of chromsome/gene type entities. At the end of my original comment, I did say that it would have to be converted into a _reference_ to satisfy specific needs.

But, in a biological sense, a _collection_ of strains makes more sense.

In any case, what I have up there is for the purpose of the upcoming discussion. The end goal is to develop a model that is comprehensive enough to support the needs of multi-organism, multi-strain, multi-genome/annotation InterMine instances.

**Comment by JoelRichardson** [comment]
@adf-ncgr - I was going to suggest exactly that (adding a strain reference to SequenceFeature). Another possibility would be to hang it off the Location object. In fact, this is where we've added (for now) the assembly version in mousemine. I could imaging the interspecific designation also being there.

**Comment by sammyjava** [comment]
Ooooh, not off of Location, please, I'm enjoying representing Location in the Neo4j graph as a relationship (edge) because it only contains a few scalar attributes. :)

**Comment by JoeCarlson** [comment]
As you know, phytozome implements multiple strains (or multiple assemblies, or whatever) as multiple entries in the organism table. This is not a "correct" implementation in any database normal form sense. What was driving us is the need to make phytozome usable to a bench biologist who wants to make a query.

Imagine someone who wants to do a query. Say, which genes have particular domain in my organism of interest. They'll type in a PFAM domain and ask for the gene and organism names. It may not be clear that in order to make sense of the results that list may come from different strains/assemblies/cultivars or whatever. In order to extract the information, they'd need to have to know to select or constrain a separate field when building the query. The critical thing is to make the results clear and allow people who may not have a complete understanding of the data model in the database make useful queries.

If the audience of your mine is different than ours - such as a breeding program - where dealing with multiple variations of one organism is the norm, then a collection of strains referencing the organism table may make sense. But I would not implement it for ourselves. We need to make the mine respond to our users' needs.

**Comment by JoelRichardson** [comment]
Hi @JoeCarlson, I see what you're saying, and I agree with designing to make things easy (as possible) the user. But as I see it, defining good templates, combined with the user's ability to filter results, goes a long way to addressing the complexity-for-the-user problem. In addition, templates provide great teaching tools/starting points for anyone who wants to roll their own. So I don't see a little normalization here as a necessary impediment to the user's ability to understand and be effective. One could make the subspecies and/or assembly versions parameters the user gets to set on a template form. Or not, and wire them into the query. If the query returns values for multiple subspecies, you could show those values in the results - which makes it obvious to the user and allows him/her to filter/sort as needed.

Also @sammyjava, to the impact storing subspecies/assembly with Location. Saying that we shouldn't do this because it makes things more complicated for neo4j, is (to me) the tail wagging the dog. The choice of database technology shouldn't drive decisions in the conceptual schema.

**Comment by adf-ncgr** [comment]
@JoelRichardson @JoeCarlson - it sounds to me as though if people aren't opposed to adding an optional Strain reference to SequenceFeature (in addition to the existing Organism reference) and a collection (potentially empty) of Strain references to Organism, then it could be left to the discretion of the Miner if they wanted to normalize out the Strain info and have Organism left to represent pure species concept (Miner Joel) or just ignore Strain in favor of coining new Organisms (Miner Joe, and Miner the "Moosher"). This might be bad for inter-intermine querying, but might be a formal solution everyone can live with?

**Comment by sammyjava** [comment]
@JoelRichardson Yeah, I was joking, but I'll be truly bummed if I have to represent Locations as nodes in the graph. It'll make it a helluva lot more cumbersome. Or, I can deviate and load them as edges anyway, references be damned. :)

**Comment by akvallejos** [comment]
In RatMine, I had started to implement strain via the OntologyTerm and Annotation objects. To my knowledge, this was never fully implemented. RGD has a strain ontology, so this made sense for us, but I am not sure how applicable it is to the rest of the community.

I am coming back from a long hiatus from InterMine. So I am still catching up on all the discussions going on.

**Comment by julie-sullivan** [comment]
**Option 1: Simple**

```
  <class name="Organism" is-interface="true">
    ... 
    <attribute name="annotationVersion" type="java.lang.String"/>
    <attribute name="assemblyVersion" type="java.lang.String"/>
    <attribute name="strainIdentifier" type="java.lang.String"/>
    <attribute name="strainName" type="java.lang.String"/>
  </class>
```
Organism name would be updated to include strain, e.g. 

```
Z. mays
Z. mays PH206
Z. mays PH207
```

**Option 2: More correct**

```
<class name="Organism" is-interface="true">
    ...
    <collection name="strains" referenced-type="Strain" reverse-reference="organism" />
</class>

<class name="Strain" is-interface="true">
    <attribute name="identifier" type="java.lang.String" />
    <attribute name="name" type="java.lang.String" />
    <reference name="organism" referenced-type="Organism" reverse-reference="strains" />
</class>

<class name="SequenceFeature" is-interface="true">
    ...
    <reference name="strain" referenced-type="Strain"/>
    <reference name="organism" referenced-type="Organism"/>
</class>
```

* Identifier + organism would no longer be enough to uniquely identify a record, so list upgrade would have to be refactored. 
* Dropdown on template / query pages would have to include both organism and strain.

**Comment by JoelRichardson** [comment]
In MouseMine, identifier+organism is still sufficient because each feature in each strain genome is uniquely identified. As well, we have separated the concept of "canonical gene" from "strain gene", and represent both: each strain gene points to its canonical gene and each canonical gene has a collection of its strain-specific forms. For example, Pax6 (MGI:97490) is a canonical gene that has a collection of 19 strain-specific genes, which includes among others MGP_AJ_G0026191, the gene in strain A/J.

**Comment by julie-sullivan** [comment]
That's a really clear way to do things.

**Comment by JoelRichardson** [comment]
This just went public, if you want to checkout it out. For example, this template: http://www.mousemine.org/mousemine/template.do?name=Gene_strain_genes&scope=global

**Comment by sammyjava** [comment]
I'm currently using Option 1, of course (calling it "variety" rather than "strain" which is a more specific terminology, but one can override how it is displayed, of course); which, as you say, makes the rest of the mine work the same. But I'd agree that an "organism" (genus+species) can contain many "strains" so Option 2 is really more correct, and I'll just have to recode a zillion classes, which is OK. So I'm fine with either one and cast a tentative vote for Option 2 as a more robust, more accurate representation.
